homeassistant/components/tfa_me/coordinator.py

```diff
@@ -0,0 +1,179 @@
+
+        # Create a single reusable TFA.me client
+        session = async_get_clientsession(hass)
+        self._client = TFAmeClient(
+            resolved_host, "sensors", log_level=1, session=session
+        )
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=self.poll_interval,
+        )
+
+    async def _async_update_data(self):
+        """Request and update data."""
+        parsed_data = {}  # dict for coordinator data
+
+        # Try to update data from station URL: e.g. "http://192.168.1.38/sensors"
+        try:
+            # Fetch all available sensors as JSON from TFA.me station
+            json_data = await self._client.async_get_sensors()
+
+            # Convert JSON to entities dict. for HA coordinator
+            parsed_data = self.json_to_entities(json_data=json_data)
+
+        # Specific error mapping
+        except (TFAmeHTTPError, TFAmeJSONError) as err:
+            # Device responding but data invalid
+            _LOGGER.exception("Invalid response from TFA.me gateway %s", self.host)
+            raise UpdateFailed(f"Invalid response: {err}") from err
+
+        except (TFAmeTimeoutError, TFAmeConnectionError, TFAmeException) as err:
+            # Timeout, connection error, other unknown client error
+            _LOGGER.exception("Error while updating TFA.me data from %s", self.host)
+            if self.first_init == 0:
+                raise ConfigEntryNotReady(f"Cannot reach {self.host}: {err}") from err
+            raise UpdateFailed(f"Connection problem: {err}") from err
+
+        except Exception as err:  # Fallback for unexpected errors
+            _LOGGER.exception(
+                "Unexpected error while updating TFA.me data from %s", self.host
+            )
+            if self.first_init == 0:
+                raise ConfigEntryNotReady(f"Unexpected error: {err}") from err
+            raise UpdateFailed(f"Unexpected error: {err}") from err
+
+        else:
+            if self.first_init < 2:
+                self.first_init += 1
+
+            # values are available at self.coordinator.data[self.entity_id]["keyword"]
+            return parsed_data
+
+    def json_to_entities(self, json_data: dict) -> Any:
+        """Convert a TFA.me JSON dictionary into a HA dictionary for coordinator."""
+        parsed_data = {}
+
+        try:
+            # Get gateway ID
+            gateway_id: str = json_data.get("gateway_id", "tfame")
+            gateway_id = gateway_id.lower()
+            self.gateway_id = gateway_id
+            # Fallback time if "timestamp" is missing
+            formatted_time_str = datetime.now().replace(microsecond=0).isoformat() + "Z"
+
+            for sensor in json_data.get("sensors", []):
+                sensor_id = sensor["sensor_id"]
+
+                for m_name, values in sensor.get("measurements", {}).items():
+                    entity_id = f"sensor.{gateway_id}_{sensor_id}_{m_name}"
```

Comment:
We shouldn't rely on this being the entity id as a user could change this. Same applies for the sensor name below.

Instead rely on the unique ID of the entity.

A better solution would be to just store the values without the knowing the context of the entity. For example:

```python
data = {
"wind_direction": value,
"rain_1_hour": value2,
}
```

Then use SensorEntityDescriptions to access the data.. For example:

https://github.com/home-assistant/core/blob/4960871c8488582d52187f25df61e73fe1994d36/homeassistant/components/nintendo_parental_controls/sensor.py#L26-L57